Identify JMRI threads with a ThreadGroup

bobjacobsen opened the pull request:
This (starts to) move toward JMRI-created threads being in the "JMRI" ThreadGroup.  This makes it easier to build tools that i.e. catch threads that leak past the end of a test.

 - Add a routine to ThreadingUtil to provide that ThreadGroup for use in Thread actors
 - Update docs
 - Convert about a dozen thread uses
 - Update the checking routines (only partially done, as not all threads migrated) in JUnitUtil.tearDown

Also 
 - Fix one issue in a test tearDown

bobjacobsen commented:
@rhwood @pabender Would appreciate your thoughts on this.  It's a bit to work to get all these threads into a common group, but I think it'll make it easier to track them in a common tearDown check for still-running threads.  (Note this doesn't fix most of the threads in #6827 because those are created outside JMRI)

bobjacobsen commented:
> I do have a little problem with adding yet another static method in the mechanisms here, I really do believe we overuse static methods. 

Any suggestions for another approach?  

We can't reasonably pass context through essentially-every method call.  (The implementation can be non-static, if needed, i.e. for different contexts)

If its just the semi-global nature that's a concern, we could do something like we do with logging and have everybody who need access instantiate an object.  Not sure what that gains if they're all instantiating the same object...

pabender commented:
> > I do have a little problem with adding yet another static method in the mechanisms here, I really do believe we overuse static methods.
> 
> Any suggestions for another approach?

Basically the two design pattern approaches to reduce the number of static calls are to either use dependency injection or to use a service loader.

So put an autodefault class that implements an auto closeable interface into our service loader (the InstanceManager) and have it pulled from there.

But again, the static calls are not something I see as a blocker for this going in as is.

Paul

rhwood commented on java/src/jmri/util/ThreadingUtil.java:
```diff
@@ -224,6 +224,25 @@ static public boolean isGUIThread() {
+     * This should be passed to as the first argument to the {@link Thread} constructor
+     * so we can track JMRI-created threads.
+     */
+    static public ThreadGroup getJmriThreadGroup() {
```
It seems to me this method can create multiple ThreadGroups named "JMRI". Are we wanting to allow that, or are we wanting there to be just one ThreadGroup named "JMRI"?

bobjacobsen replied:
Once the "JMRI" group has been created on the first pass, this navigates the ThreadGroup tree to find the guaranteed "main" root, then down to find the "JMRI" group.

bobjacobsen commented:
I'll add the `newThread` methods. 

There's about 40 places where the JMRI class extends ThreadGroup. They'll have to be refactored to use newThread.

rhwood commented:
> There's about 40 places where the JMRI class extends ThreadGroup. They'll have to be refactored to use newThread.

We can use a mix of `new Thread...` and `ThreadingUtil.newThread...`; I’m not stating the former shouldn’t be usable, just that we might want to provide a shortcut to encourage the use of the ThreadGroup.

bobjacobsen commented:
Added the `newThread` helpers, switched over most of the `jmrit` threads.

Encountered a security exception on Travis _only_:
```
Exception in thread "dns.close in ZeroConfServiceManager#stopAll" Exception in thread "dns.close in ZeroConfServiceManager#stopAll" java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "modifyThread")
	at java.security.AccessControlContext.checkPermission(AccessControlContext.java:472)
	at java.security.AccessController.checkPermission(AccessController.java:884)
	at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
	at java.util.concurrent.ThreadPoolExecutor.checkShutdownAccess(ThreadPoolExecutor.java:742)
	at java.util.concurrent.ThreadPoolExecutor.shutdown(ThreadPoolExecutor.java:1397)
	at java.util.concurrent.Executors$DelegatedExecutorService.shutdown(Executors.java:669)
	at javax.jmdns.impl.JmDNSImpl.close(JmDNSImpl.java:1935)
	at jmri.util.zeroconf.ZeroConfServiceManager.lambda$null$5(ZeroConfServiceManager.java:315)
	at java.lang.Thread.run(Thread.java:748)
```

Turns out this is not a Travis-AppVeyor-macOS difference, but actually a Maven vs Ant vs sh difference compounded with what's apparently a problem in `javax.jmdns.impl.JmDNSImpl.close`.  That routine is (apparently) trying to shut down a system service, which involves shutting down system threads.  Our security.policy file doesn't enable that (and it would be hard to make it do that). Ant and the sh launcher don't have the `-Djava.security.manager policy` argument that enables the security manager; that seems to be for historical reasons.  `pom.xml` had it, so was tripping the issue.  Perhaps the jmdns problem can be fixed, in which case we can turn the security manager back on, but I removed the `-Djava.security.manager policy` from pom.xml so there's common behavior.

bobjacobsen commented:
Travis graphical

please don't remove [main] jmri.util.JUnitAppenderTest.testUnexpectedCheck()
852[ERROR] Errors: 
853[ERROR]   SRCPMonPaneTest>AbstractMonPaneTestBase.testFreezeButton:119 » TimeoutExpired ...
854[INFO] 
855[ERROR] Tests run: 25755, Failures: 0, Errors: 1, Skipped: 3155

bobjacobsen commented:
Travis graphical
AssertTest: asserts are enabled
853[ERROR] Failures: 
854[ERROR]   PrintTrainBuildReportActionTest.testPrintAction Multiple Failures (2 failures)
855	java.lang.AssertionError: "wait to complete" did not occur in time
856	java.lang.AssertionError: Shutdown task found: Save Operations State
857[INFO] 
858[ERROR] Tests run: 25759, Failures: 1, Errors: 0, Skipped: 3155
